From a59c523947ed4b5996e19f60370fd1b2f6e75e71 Mon Sep 17 00:00:00 2001
From: Daniel Baluta <daniel.baluta@nxp.com>
Date: Wed, 3 Oct 2018 16:45:57 +0300
Subject: [PATCH 4844/5242] MLK-18497-13: ASoC: fsl: dsp: Skip SDRAM section
 update if fw is already loaded

commit  a0cffd9a9299362f2b3a4d24b78a8574b736fdfa from
https://source.codeaurora.org/external/imx/linux-imx.git

If the DSP firmware binary is already loaded it is wrong to update
SDRAM located sections because we will overwrite and data stored there.

This makes suspend/resume work.

Reviewed-by: Cosmin-Gabriel Samoila <cosmin.samoila@nxp.com>
Signed-off-by: Shengjiu Wang <shengjiu.wang@nxp.com
Signed-off-by: Daniel Baluta <daniel.baluta@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/fsl/fsl_dsp.c       |   12 ++++++++----
 sound/soc/fsl/fsl_dsp_proxy.h |    1 +
 2 files changed, 9 insertions(+), 4 deletions(-)

diff --git a/sound/soc/fsl/fsl_dsp.c b/sound/soc/fsl/fsl_dsp.c
index 432e179..3ea7e75 100644
--- a/sound/soc/fsl/fsl_dsp.c
+++ b/sound/soc/fsl/fsl_dsp.c
@@ -369,6 +369,7 @@ int fsl_dsp_open_func(struct fsl_dsp *dsp_priv, struct xf_client *client)
 	atomic_set(&client->vm_use, 0);
 
 	client->global = (void *)dsp_priv;
+	dsp_priv->proxy.is_loaded = 0;
 
 	pm_runtime_get_sync(dev);
 
@@ -688,10 +689,12 @@ static void dsp_load_firmware(const struct firmware *fw, void *context)
 				(!strcmp(&strtab[shdr->sh_name], ".data"))   ||
 				(!strcmp(&strtab[shdr->sh_name], ".bss"))
 			) {
-				memcpy_dsp((void *)(dsp_priv->sdram_vir_addr
-				  + (sh_addr - dsp_priv->sdram_phys_addr)),
-				  (const void *)image,
-				  shdr->sh_size);
+				if (!dsp_priv->proxy.is_loaded) {
+					memcpy_dsp((void *)(dsp_priv->sdram_vir_addr
+				  	+ (sh_addr - dsp_priv->sdram_phys_addr)),
+				  	(const void *)image,
+				  	shdr->sh_size);
+				}
 			} else {
 				/* sh_addr is from DSP view, we need to
 				 * fixup addr because we load the firmware from
@@ -710,6 +713,7 @@ static void dsp_load_firmware(const struct firmware *fw, void *context)
 	/* start the core */
 	sc_pm_cpu_start(dsp_priv->dsp_ipcHandle,
 					SC_R_DSP, true, dsp_priv->iram);
+	dsp_priv->proxy.is_loaded = 1;
 }
 
 /* Initialization of the MU code. */
diff --git a/sound/soc/fsl/fsl_dsp_proxy.h b/sound/soc/fsl/fsl_dsp_proxy.h
index fabcf76..8df703e 100644
--- a/sound/soc/fsl/fsl_dsp_proxy.h
+++ b/sound/soc/fsl/fsl_dsp_proxy.h
@@ -299,6 +299,7 @@ struct xf_proxy {
 
 	struct completion	cmd_complete;
 	int			is_ready;
+	int			is_loaded;
 
 	/* ...internal lock */
 	spinlock_t              lock;
-- 
1.7.9.5

